dev/doc/proof-engine.md

```diff
@@ -0,0 +1,130 @@
+Tutorial on the new proof engine for ML tactic writers
+======================================================
+
+Starting from Coq 8.5, a new proof engine has been introduced, replacing the old
+meta-based engine which had a lot of both expressivity and soundness drawbacks.
```

Comment:
I think the most important change from the point of view of this tutorial is that the new proof engine uses an opaque type, so as to allow changing the underlying representation.

This permits some gains both in expressiveness and soundness. But they are not that huge, all things considered:

- Dependent subgoals (the original motivation)
- Multiple-goal tactics
- More consistent treatment of evar updates
- Avoid bug duplication by implementing one more type-checking algorithm (in particular several bug of `refine` closed)
- Stuff about metas
- Easier management of evar_maps

All of these were problems. But not major ones (not to everybody at any rate: at least some of them were for me, obviously). So maybe emphasising the gain in flexibility in the code would make for a better _entrée en matière_.